sockeye/data_io.py

```diff
@@ -495,211 +1215,142 @@ def __init__(self,
-            self.data_target[i] = np.asarray(self.data_target[i], dtype=self.dtype)
-            self.data_label[i] = np.asarray(self.data_label[i], dtype=self.dtype)
-
-            n = len(self.data_source[i])
-            batch_size_seq = self.bucket_batch_sizes[i].batch_size
-            if n % batch_size_seq != 0:
-                buck_shape = self.buckets[i]
-                rest = batch_size_seq - n % batch_size_seq
-                if self.fill_up == 'pad':
-                    raise NotImplementedError
-                elif self.fill_up == 'replicate':
-                    logger.info("Replicating %d random sentences from bucket %s to size it to multiple of %d", rest,
-                                buck_shape, batch_size_seq)
-                    random_indices = np.random.randint(self.data_source[i].shape[0], size=rest)
-                    self.data_source[i] = np.concatenate((self.data_source[i], self.data_source[i][random_indices, :]),
-                                                         axis=0)
-                    self.data_target[i] = np.concatenate((self.data_target[i], self.data_target[i][random_indices, :]),
-                                                         axis=0)
-                    self.data_label[i] = np.concatenate((self.data_label[i], self.data_label[i][random_indices, :]),
-                                                         axis=0)
+                raise StopIteration
+        return self.shard_iter.next()
+
+    def save_state(self, fname: str):
+        with open(fname, "wb") as fp:
+            pickle.dump(self.shards_fnames, fp)
+            pickle.dump(self.shard_index, fp)
+        self.shard_iter.save_state(fname + ".sharditer")
+
+    def load_state(self, fname: str):
+        with open(fname, "rb") as fp:
+            self.shards_fnames = pickle.load(fp)
+            self.shard_index = pickle.load(fp)
+        self._load_shard()
+        self.shard_iter.load_state(fname + ".sharditer")
+
+
+class ParallelSampleIter(BaseParallelSampleIter):
+
+    def __init__(self,
+                 data: ParallelDataSet,
+                 buckets,
+                 batch_size,
+                 bucket_batch_sizes,
+                 source_data_name=C.SOURCE_NAME,
+                 target_data_name=C.TARGET_NAME,
+                 label_name=C.TARGET_LABEL_NAME,
+                 dtype='float32') -> None:
+        super().__init__(buckets, batch_size, bucket_batch_sizes,
+                         source_data_name, target_data_name, label_name, dtype)
+
+        # create independent lists to be shuffled
+        self.data = ParallelDataSet(list(data.source), list(data.target), list(data.label))
+
+        # create index tuples (buck_idx, batch_start_pos) into buckets. These will be shuffled.
+        self.batch_indices = get_batch_indices(self.data, bucket_batch_sizes)
+        self.curr_batch_index = 0
+
+        self.inverse_data_permutations = [mx.nd.arange(0, max(1, self.data.source[i].shape[0]))
```

Comment:
We can get rid of the type annotations (which are not corect anyway)

Reply:
done